react/src/components/PostHogFeature.tsx

```diff
@@ -0,0 +1,94 @@
+import { useFeatureFlagPayload, useFeatureFlagVariantKey, usePostHog } from '../hooks'
+import React, { RefObject, useEffect, useMemo, useRef, useState } from 'react'
+import { PostHog } from '../context'
+
+export type PostHogFeatureProps = {
+    flag: string
+    children: React.ReactNode | ((payload: any) => React.ReactNode)
+    match?: string | boolean
+    visibilityObserverOptions?: IntersectionObserverInit
+}
+
+export function PostHogFeature({
+    flag,
+    match,
+    children,
+    visibilityObserverOptions,
+}: PostHogFeatureProps): JSX.Element | null {
+    const posthog = usePostHog()
+    const payload = useFeatureFlagPayload(flag)
+    const variant = useFeatureFlagVariantKey(flag)
+    const [clickTracked, setclickTracked] = useState(false)
+
+    if (match === undefined || variant === match) {
+        const childNode: React.ReactNode = typeof children === 'function' ? children(payload) : children
+        return (
+            <div
+                onClick={() => {
+                    if (!clickTracked) {
+                        trackClicks(flag, posthog)
+                        setclickTracked(true)
+                    }
+                }}
+            >
+                <VisibilityTracker flag={flag} options={visibilityObserverOptions}>
+                    {childNode}
+                </VisibilityTracker>
+            </div>
+        )
+    }
+
+    return null
+}
+
+function trackClicks(flag: string, posthog?: PostHog) {
+    posthog?.capture('$feature_flag_clicked', { feature_flag: flag, $set: { [`$feature_interaction/${flag}`]: true } })
+}
+
+function trackVisibility(flag: string, posthog?: PostHog) {
+    posthog?.capture('$feature_flag_viewed', { feature_flag: flag })
+}
+
+function VisibilityTracker({
+    flag,
+    children,
+    options,
+}: {
+    flag: string
+    children: React.ReactNode
+    options?: IntersectionObserverInit
+}): JSX.Element {
+    const ref = useRef<HTMLDivElement>(null)
+    const posthog = usePostHog()
+    const [tracked, setTracked] = useState(false)
+
+    const isIntersecting = useVisibleOnScreen(ref, {
+        threshold: 0.1,
+        ...options,
+    })
+
+    if (isIntersecting && !tracked) {
+        trackVisibility(flag, posthog)
+        setTracked(true)
+    }
```

Comment:
... Yes, I spent a while trying to parse this but I'm not sure what this means.

Great comments overall, thanks for the review! 🙌

Reply:
I think I get it now, this is indeed dumb, I should be using a single-use `useEffect` instead to run it once, rather than re-rendering the component because the component state isn't changing. Will fix 🤦 .

Reply:
.. oh, but, these functions need to run when isIntersecting happens for the first time only - I don't get how you'll model that without state here?

Reply:
.. or did you mean merge isTracked & tracked into `useVisibleOnScreen` ?

Reply:
Refs 😊 you're only using the state to make sure you don't fire more than once - if you're not changing something in the rendered content then we don't need state. A simple ref would do